### PR TITLE
Issue #3630 Forwarded-Port

### PR DESCRIPTION
**Pull request by gregw**
#3630
Added support for the X-Forwarded-Port header.
HostPort now handles the case of a stand alone IPv6 address
Reimplemented header scanning using more efficient Trie and MethodHandles

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
Looking at this, it could be a lot more efficient and since it runs on every request it is well worth the effort.... so stand by for some more commits

File: jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java [hunk 1]
```diff
@@ -41,7 +41,8 @@
                 Arguments.of("10.10.10.1", "10.10.10.1", null),
                 Arguments.of("10.10.10.1:80", "10.10.10.1", "80"),
                 Arguments.of("[0::0::0::1]", "[0::0::0::1]", null),
-                Arguments.of("[0::0::0::1]:80", "[0::0::0::1]", "80")
+                Arguments.of("[0::0::0::1]:80", "[0::0::0::1]", "80"),
+                Arguments.of("0:1:2:3:4:5:6","[0:1:2:3:4:5:6]",null)
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java]: OK.

**Comment by gregw** [comment]
I have added a benchmark of the new Trie/MethodHandle approach.   It is about 40% faster, but does produce some garbage... but at least nothing that goes old gen:
```
Benchmark                                                            Mode  Cnt         Score         Error   Units
ForwardBenchmark.testStringCompare                                  thrpt   10  20467765.674 ±  246064.823   ops/s
ForwardBenchmark.testStringCompare:·gc.alloc.rate                   thrpt   10         0.032 ±       0.002  MB/sec
ForwardBenchmark.testStringCompare:·gc.alloc.rate.norm              thrpt   10         0.003 ±       0.001    B/op
ForwardBenchmark.testStringCompare:·gc.count                        thrpt   10           ≈ 0                counts
ForwardBenchmark.testTrieMethodHandle                               thrpt   10  28399428.722 ± 2404195.291   ops/s
ForwardBenchmark.testTrieMethodHandle:·gc.alloc.rate                thrpt   10       261.374 ±      19.214  MB/sec
ForwardBenchmark.testTrieMethodHandle:·gc.alloc.rate.norm           thrpt   10        16.002 ±       0.001    B/op
ForwardBenchmark.testTrieMethodHandle:·gc.churn.G1_Eden_Space       thrpt   10       263.723 ±     159.580  MB/sec
ForwardBenchmark.testTrieMethodHandle:·gc.churn.G1_Eden_Space.norm  thrpt   10        16.267 ±      10.286    B/op
ForwardBenchmark.testTrieMethodHandle:·gc.churn.G1_Old_Gen          thrpt   10        ≈ 10⁻⁴                MB/sec
ForwardBenchmark.testTrieMethodHandle:·gc.churn.G1_Old_Gen.norm     thrpt   10        ≈ 10⁻⁵                  B/op
ForwardBenchmark.testTrieMethodHandle:·gc.count                     thrpt   10        13.000                counts
ForwardBenchmark.testTrieMethodHandle:·gc.time                      thrpt   10        18.000                    ms
```

So whilst I don't think this was ever a lot of the time spent in the server, as the customizer is run on every request, it is a good thing to optimise.

**Comment by joakime** [comment]
> So whilst I don't think this was ever a lot of the time spent in the server, as the customizer is run on every request, it is a good thing to optimise.

This PR does more, and is 10% faster, I'd call that "a good thing" :tm:.

**Pull request merged**